db.adoc: Refined the 'Configuring a database' paragraph (#32483)
keycloak/keycloak: docs/guides/server/db.adoc

For each supported database, the server provides some opinionated defaults to simplify database configuration. You complete the configuration by providing some key settings such as the database host and credentials.

> The configuration can be set during a *build command* OR a *start command*:

Reviewer: Should it be mentioned that other non-optimized commands (import, export, bootstrap-admin) can configure the database as well?

Author: I'll mention import and export, but bootstrap-admin seems unknown to kc.sh.

Reviewer: It's new to version 26

. Using a *build command* followed by an optimzed start command (recommended)
+
----

> bin/kc.[sh|bat] build --db postgres --db-url-host mypostgres --db-username myuser --db-password change_me

Reviewer: I know it's being done for simplicity, but should we call out that specifying the password via a cli option is not recommended?

Author: cf. the conf/keycloak.conf comment above for the build command.
since the build command is recommended, it's already specified

> bin/kc.[sh|bat] build --db postgres --db-url-host mypostgres --db-username myuser --db-password change_me
> bin/kc.[sh|bat] start --optimized

Reviewer: Only the build time options can be specified on the build command

```suggestion
bin/kc.[sh|bat] build --db postgres 
bin/kc.[sh|bat] start --optimized --db-url-host mypostgres --db-username myuser --db-password change_me
```

Author: Ok, I missed that detail. The whole first point is to configure the database using only the build command. I would suggest to use the conf/keycloak.conf file instead to avoid having settings spread across 2 different commands.

Reviewer: > Ok, I missed that detail. The whole first point is to configure the database using only the build command.

Unfortunately we're dealing with both build time and run time options. You configure build time database type with the build command, then the database related runtime options on the start command.

> I would suggest to use the conf/keycloak.conf file instead to avoid having settings spread across 2 different commands.

Right, most will use that or environment variables when creating Docker images - calling that out in general (not specifically for the optimized approach) will also address the concern about the password being shown directly on the cli.

----
+

These commands include the minimum settings needed to connect to the database.

. Using *only a start command* (without --optimized)
+
<@kc.start parameters="--db postgres --db-url-host mypostgres --db-username myuser --db-password change_me"/>
+

This command includes the minimum settings needed to connect to the database.

> The default schema is `keycloak`, but you can change it by using the `--db-schema` configuration option.

Reviewer: ```suggestion
The default schema is `keycloak`, but you can change it by using the `db-schema` configuration option.
```

Since this can be set in other ways, so it doesn't need to be specific to the cli option.

Author: OK

For each supported database, the server provides some opinionated defaults to simplify database configuration. You complete the configuration by providing some key settings such as the database host and credentials.

The configuration can be set during a *build command* OR a *start command*:

. Using a *build command* followed by an optimzed start command (recommended)
+

> First, the minimum settings needed to connect to the database can be specified in *conf/keycloak.conf*:

Reviewer 2: ```suggestion
First, the minimum settings needed to connect to the database can be specified in `conf/keycloak.conf`:
```

+
----
# The database vendor.
db=postgres

# The username of the database user.
db-username=keycloak

# The password of the database user.

> db-password=change_me

Reviewer 3: Yeah, we should probably added some warnings to other places too.

For this one, I added a suggestion how to make it a bit more prominent. Let me know what you think.

# Sets the hostname of the default JDBC URL of the chosen vendor
db-url-host=mypostgres
----
+

Then, the following commands create a new and optimized server image based on the configuration options and start the server.
+
----
bin/kc.[sh|bat] build
bin/kc.[sh|bat] start --optimized
----
+

. Using *only a start command* (without --optimized)
+
<@kc.start parameters="--db postgres --db-url-host mypostgres --db-username myuser --db-password change_me"/>
+

> This command includes the minimum settings needed to connect to the database but it exposes the database password and is not recommended.

Reviewer: ```suggestion
This command includes the minimum settings needed to connect to the database but it exposes the database password and is not recommended. Use the `conf/keycloak.conf` as shown above, environment variables, or keystore for at least the password.
```

> <@kc.start parameters="--db postgres --db-url-host keycloak-postgres --db-username keycloak --db-password change_me"/>
> +
> 
> This command includes the minimum settings needed to connect to the database but it exposes the database password and is not recommended. Use the `conf/keycloak.conf` as shown above, environment variables, or keystore for at least the password.

Reviewer 3: ```suggestion
<@kc.start parameters="--db postgres --db-url-host keycloak-postgres --db-username keycloak --db-password change_me"/>

WARNING: The examples above include the minimum settings needed to connect to the database but it exposes the database password and is not recommended. Use the `conf/keycloak.conf` as shown above, environment variables, or keystore for at least the password.
```

The default schema is `keycloak`, but you can change it by using the `db-schema` configuration option.

It is also possible to configure the database when https://www.keycloak.org/server/importExport[importing or exporting Realms] or https://www.keycloak.org/server/bootstrap-admin-recovery[creating new Keycloak admin users / service accounts] (starting with release 26):
----
bin/kc.[sh|bat] import --help
bin/kc.[sh|bat] export --help
bin/kc.[sh|bat] bootstrap-admin --help
----

> For more information, see <@links.server id="configuration"/>.

Reviewer: The example shown in https://www.keycloak.org/server/configuration is nearly identical to what is shown here. Seems like this could be aligned further.

> It is also possible to configure the database when https://www.keycloak.org/server/importExport[importing or exporting Realms] or https://www.keycloak.org/server/bootstrap-admin-recovery[creating new Keycloak admin users / service accounts] (starting with release 26):

Reviewer: ```suggestion
It is also possible to configure the database when <@links.server id="importExport"/> or <@links.server id="bootstrap-admin-recovery"/>:
```

Author: without mentioning '(starting with release 26)'?

Reviewer: > without mentioning '(starting with release 26)'?

Yes, we don't typically do that as the docs are versioned as well.
